<a href="https://colab.research.google.com/github/Sam-krish2411/DATA-SCIENCE-ASSIGNMENT/blob/main/ASSOCIATION_RULES.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [6]:
import warnings
warnings.filterwarnings("ignore")

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [7]:
!pip install mlxtend

# Import libraries
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules


In [8]:
#Load Dataset

df = pd.read_excel("Online retail.xlsx", header=None)

# Convert the column into a list of transactions
transactions = df.iloc[:, 0].dropna().tolist()
transactions = [t.split(",") for t in transactions]

In [9]:
# 2. One-Hot Encode
# -------------------------
te = TransactionEncoder()
te_ary = te.fit(transactions).transform(transactions)
df_encoded = pd.DataFrame(te_ary, columns=te.columns_)

print("One-hot encoded dataset:")
print(df_encoded.head())

One-hot encoded dataset:
    asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0       False     True               True      False     True        False   
1       False    False              False      False    False        False   
2       False    False              False      False    False        False   
3       False    False              False      False     True        False   
4       False    False              False      False    False        False   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0  False           False      False        False  ...   False            True   
1  False           False      False        False  ...   False           False   
2  False           False      False        False  ...   False           False   
3  False           False      False        False  ...    True           False   
4  False           False      False        False  ...   False           False   

   water spray  whi

In [11]:
# Apply Apriori

frequent_itemsets = apriori(df_encoded, min_support=0.05, use_colnames=True)
frequent_itemsets = frequent_itemsets.sort_values(by="support", ascending=False)

print("\nFrequent Itemsets:")
print(frequent_itemsets.head(10))


Frequent Itemsets:
     support             itemsets
16  0.238368      (mineral water)
6   0.179709               (eggs)
21  0.174110          (spaghetti)
8   0.170911       (french fries)
3   0.163845          (chocolate)
12  0.132116          (green tea)
15  0.129583               (milk)
13  0.098254        (ground beef)
10  0.095321  (frozen vegetables)
18  0.095054           (pancakes)


In [12]:
# Generate Association Rules

rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1.0)
rules = rules.sort_values(by="lift", ascending=False)

print("\nTop Association Rules:")
print(rules[['antecedents', 'consequents', 'support', 'confidence', 'lift']].head(10))


Top Association Rules:
       antecedents      consequents   support  confidence      lift
0  (mineral water)      (spaghetti)  0.059725    0.250559  1.439085
1      (spaghetti)  (mineral water)  0.059725    0.343032  1.439085
3      (chocolate)  (mineral water)  0.052660    0.321400  1.348332
2  (mineral water)      (chocolate)  0.052660    0.220917  1.348332
4  (mineral water)           (eggs)  0.050927    0.213647  1.188845
5           (eggs)  (mineral water)  0.050927    0.283383  1.188845


### Interpretation of Top Association Rules

The following rules reveal interesting co-purchase patterns in the retail dataset:

1. **Mineral Water → Spaghetti**
   - **Support**: 5.97% of transactions contain both items.
   - **Confidence**: 25.05% of transactions with mineral water also include spaghetti.
   - **Lift**: 1.44 — customers who buy mineral water are **44% more likely** to buy spaghetti compared to random chance.

2. **Spaghetti → Mineral Water**
   - **Confidence**: 34.30% — stronger than the reverse rule, indicating spaghetti buyers are more likely to also buy mineral water.
   - **Lift**: 1.44 — confirms a strong positive association.

3. **Chocolate → Mineral Water**
   - **Confidence**: 32.14% — nearly one-third of chocolate buyers also purchase mineral water.
   - **Lift**: 1.35 — suggests a moderate association.

4. **Mineral Water → Chocolate**
   - **Confidence**: 22.09% — weaker than the reverse rule, but still notable.
   - **Lift**: 1.35 — consistent with the chocolate-mineral water pairing.

5. **Mineral Water → Eggs**
   - **Confidence**: 21.36% — about one-fifth of mineral water buyers also buy eggs.
   - **Lift**: 1.19 — mild association.

6. **Eggs → Mineral Water**
   - **Confidence**: 28.34% — stronger than the reverse, indicating eggs buyers often include mineral water.
   - **Lift**: 1.19 — confirms a weak but positive link.

### Insights
- **Mineral water** appears frequently as a co-purchased item, acting as a common companion across multiple products.
- **Spaghetti and chocolate** show strong bidirectional associations with mineral water, suggesting bundled or habitual purchases.
- These rules can inform **product placement**, **cross-promotions**, or **bundle recommendations** in an online retail setting.

### 📌 Interview Questions — Association Rule Mining

#### 1. What is Lift and why is it important in Association Rules?

- **Lift** measures how much more likely two items are to be purchased together compared to if they were independent.
- It is calculated as:  
  \[
  \text{Lift}(A → B) = \frac{\text{Confidence}(A → B)}{\text{Support}(B)}
  \]
- **Interpretation**:
  - **Lift > 1**: Positive association — items co-occur more than expected.
  - **Lift = 1**: No association — items are independent.
  - **Lift < 1**: Negative association — items co-occur less than expected.
- **Importance**: Lift helps filter out rules that are statistically significant and not just frequent. It highlights **true dependencies** between items.
---
#### 2. What is Support and Confidence? How do you calculate them?

- **Support**: Proportion of transactions that contain the itemset.
  - Formula:  
    \[
    \text{Support}(A → B) = \frac{\text{Transactions containing both A and B}}{\text{Total transactions}}
    \]
- **Confidence**: Likelihood that item B is purchased when item A is purchased.
  - Formula:  
    \[
    \text{Confidence}(A → B) = \frac{\text{Support}(A ∪ B)}{\text{Support}(A)}
    \]
- **Use**:
  - Support filters out rare combinations.
  - Confidence evaluates the strength of implication.
---
#### 3. What are some limitations or challenges of Association Rule Mining?

- **Sparse data**: Many item combinations may occur infrequently, leading to low support.
- **Rule explosion**: Large datasets can generate thousands of rules, many of which are trivial or redundant.
- **No causality**: Association rules show correlation, not causation.
- **Parameter tuning**: Choosing thresholds for support, confidence, and lift requires domain knowledge.
- **Scalability**: Algorithms like Apriori can be computationally expensive for large datasets.
- **Interpretability**: Not all high-confidence rules are meaningful or actionable.
